Add ECE support for multiple product types

asumaran opened the pull request:
## Changes proposed in this Pull Request

Closes https://github.com/Automattic/woocommerce-payments/issues/8871
Fixes https://github.com/Automattic/woocommerce-payments/issues/8981


Add support for the same product types that we currently support with PRBs.

## Testing Instructions

> [!IMPORTANT]  
> I recommend skipping the failed test cases and not wasting time reviewing them, as they fail due to limitations in ECE and PRBs. For more information, refer to issues #9020 and #8373.

> [!IMPORTANT]  
> Due to the amount of testing cases I only tested it on Chrome with Google Pay using my own credit cards. **Ensure your site is in test mode**.

> [!IMPORTANT]  
> I'd recommend clearing cookies before running each test case to catch any possible errors if some data is missing.




**Setup**

- Pull and switch to the branch `as-8871-ece-on-supported-product-types`.
- Enable ECE by running `wp option update _wcpay_feature_stripe_ece 1` in your merchant's root directory.
- If ECE doesn't render, ensure your site is running over HTTPS.

To alternate how the tax is calculated:

- Go to WooCommerce > Settings to enable Taxes.
- Go to WooCommerce > Settings > Tax. Change the "Calculate tax based on" option according to the test case.

### Product Type: `simple`

<details>
<summary>Testing steps</summary>

In order to test this product type you'll need to create:

- A simple product.
- A simple virtual product.


#### Product Page

- Virtual Product
  - [x] Without Tax.
  - [ ] ❌ With Tax (Customer Shipping Address) – **Virtual Product With Tax Error (Shipping)**
  - [ ] ❌ With Tax (Customer Billing Address) – **Virtual Product With Tax Error (Billing)**
- Physical Product
  - [x] Without Tax
  - [x] With Tax (Customer Shipping Address)
  - [ ] ❌ With Tax (Customer Billing Address) – Check **Physical Product With Tax Error (Billing)**

Also test in the Cart Shortcode Page, Checkout shortcode Page, Cart Block and Checkout Block.

</details>

### Product Type: `variable`

This product is the parent of variations. It cannot be purchased; only its variations can be bought. Ensuring the ECE button renders properly on the product page is sufficient to verify that ECE is compatible with this product type.

### Product Type: `variation`

<details>
<summary>Test cases</summary>

In order to test this product type you'll need to create the following products:

- A variable product with virtual variations
  - Create a variable product.
  - Add an attribute with two values. e.g. `color`, `red|green`. Check the "Used for variations" options.
  - Create a variation for each attribute.

- A variable product with physical (simple products) variations.
  - Create a variable product.
  - Add an attribute with two values. e.g. `color`, `red|green`.
  - Create a variation for each attribute.

- A variable product with mixed products (simple and virtual) variations. 
  - Create a variable product.
  - Add an attribute with two values. e.g. `color`, `red|green`.
  - Create a variation for each attribute.
  - Set one variation as "virtual".


**Product Page**

- Virtual Products
  - [x] Without Tax
  - [ ] ❌ With Tax (Customer Shipping Address) - **Virtual Product With Tax Error (Shipping Address)**
  - [ ] ❌ With Tax (Customer Billing Address) - **Virtual Product With Tax Error (Billing Address)**
- Physical Products
  - [x] Without Tax
  - [x] With Tax (Customer Shipping Address)
  - [ ] ❌ With Tax (Customer Billing Address) – Check **Physical Product With Tax Error (Billing)**
- Mixed Products (Virtual & Physical)
  - Virtual Variation
    - [x] Without Tax
    - [ ] ❌ With Tax (Customer Shipping Address) - **Virtual Product With Tax Error 
    - [ ] ❌ With Tax (Customer Billing Address) – Check **Physical Product With Tax Error 
(Shipping Address)**
  - Physical Variation
    - [x] Without Tax
    - [x] With Tax (Customer Shipping Address)
    - [ ] ❌ With Tax (Customer Billing Address) – Check **Physical Product With Tax Error
</details>

### Product Type: `subscription`

<details>
<summary>Test cases</summary>

In order to test this product type you'll need to download and install the "WooCommerce Subscriptions" extension from https://woocommerce.com/my-account/downloads/.

Create two simple subscriptions:

- A virtual subscription.
- A simple subscription.

**Product Page**

- Virtual
  - [x] Without Tax
  - [ ] ❌ With Tax (Customer Shipping Address) - **Virtual Product With Tax Error (Shipping Address)**
  - [ ] ❌ With Tax (Customer Billing Address) - **Virtual Product With Tax Error**
- Physical
  - [x] Without Tax
  - [x] With Tax (Customer Shipping Address)
  - [ ] ❌ With Tax (Customer Billing Address) – Check **Physical Product With Tax Error**
</details>

### Product type: `variable-subscription`

This product is the parent of subscription variations. Ensuring the ECE button renders properly on the product page of a variable subscription is sufficient to verify that ECE is compatible with this product type.

### Product type: `subscription_variation`

<details>
<summary>Test cases</summary>

In order to test this product type you'll need to install the [WooCommerce Subscriptions](https://woocommerce.com/products/woocommerce-subscriptions/) plugin. You can download it from [here](https://woocommerce.com/my-account/downloads/).


**Create A Variable subscription with virtual variations**
- Create a Variable subscription.
- Add an attribute with two values. e.g. `size`, `small|big`. Check the "Used for variations" options.
- Create a variation for each attribute.

**A Variable subscription with simple variations.**
- Create a Variable subscription.
- Add an attribute with two values. e.g. `size`, `small|big`.
- Create a variation for each attribute.

**A variable product with mixed products (simple and virtual) variations.**
- Create a variable product.
- Add an attribute with two values. e.g. `size`, `small|big`.
- Create a variation for each attribute.
- Set one variation as "virtual".

**Product Page**

- Virtual
  - [x] Without Tax
  - [ ] ❌ With Tax (Customer Shipping Address) - **Virtual Product With Tax Error (Shipping Address)**
  - [ ] ❌ With Tax (Customer Billing Address) - **Virtual Product With Tax Error (Billing Address)**
- Physical
  - [x] Without Tax
  - [x] With Tax (Customer Shipping Address)
  - [ ] ❌ With Tax (Customer Billing Address) – Check **Physical Product With Tax Error (Billing)**
- Mixed Products (Virtual & Physical)
  - Virtual Variation
    - [x] Without Tax
    - [ ] ❌ With Tax (Customer Shipping Address) - **Virtual Product With Tax Error**
    - [ ] ❌ With Tax (Customer Billing Address) – **Virtual Product With Tax Error**
(Shipping Address)**
  - Physical Variation
    - [x] Without Tax
    - [x] With Tax (Customer Shipping Address)
    - [ ] ❌ With Tax (Customer Billing Address) – Check **Physical Product With Tax Error
</details>

### Product type: `booking`

<details>
<summary>Testing steps</summary>

In order to test this product type you'll need to install the [WooCommerce Bookings](https://woocommerce.com/products/woocommerce-bookings/) plugin. You can download it from [here](https://woocommerce.com/my-account/downloads/).

**Create a simple booking**

- Go to Products > Add New.
- Select "Bookable" from the product data section.
- Set the cost in the "Costs" tab.
- Publish.

**Create a virtual booking**

- Go to Products > Add New.
- Select "Bookable" from the product data section.
- Mark the "Virtual" checkbox.
- Set the cost in the "Costs" tab.
- Publish.

**Note:** No need to test the ECE on the product page for this product type as it's currently not supported in that page.

**Test cases**

- Simple Booking
  - Shortcode - Cart
    - [x] Without Tax
    - [x] With Tax (Customer Shipping Address)
    - [ ] ❌ With Tax (Customer Billing Address) – **Physical Product With Tax Error** 
  - Shortcode - Checkout
    - [x] Without Tax
    - [x] With Tax (Customer Shipping Address)
    - [ ] ❌ With Tax (Customer Billing Address) – **Physical Product With Tax Error**
  - Blocks - Cart
    - [x] Without Tax
    - [x] With Tax (Customer Shipping Address)
    - [ ] ❌ With Tax (Customer Billing Address) – **Physical Product With Tax Error**
  - Blocks - Checkout
    - [x] Without Tax
    - [x] With Tax (Customer Shipping Address)
    - [ ] ❌ With Tax (Customer Billing Address) – **Physical Product With Tax Error**
- Virtual Booking
  - Shortcode - Cart
    - [x] Without Tax
    - [ ] ❌ With Tax (Customer Shipping Address) - **Virtual Product With Tax Error**
    - [ ] ❌ With Tax (Customer Billing Address) – **Virtual Product With Tax Error**
  - Shortcode - Checkout
    - [x] Without Tax
    - [ ] ❌ With Tax (Customer Shipping Address) - **Virtual Product With Tax Error**
    - [ ] ❌ With Tax (Customer Billing Address) – **Virtual Product With Tax Error**
  - Blocks - Cart
    - [x] Without Tax
    - [ ] ❌ With Tax (Customer Shipping Address) - **Virtual Product With Tax Error**
    - [ ] ❌ With Tax (Customer Billing Address) – **Virtual Product With Tax Error**
  - Blocks - Checkout
    - [x] Without Tax
    - [ ] ❌ With Tax (Customer Shipping Address) - **Virtual Product With Tax Error**
    - [ ] ❌ With Tax (Customer Billing Address) – **Virtual Product With Tax Error**
</details>

### Product type: `bundles`

<details>
<summary>Testing steps</summary>

In order to test this product type you'll need to install the [Product Bundles](https://woocommerce.com/products/product-bundles/) plugin. You can download it from [here](https://woocommerce.com/my-account/downloads/).

**Create a bundle with simple products**

- Go to Products > Add New.
- Select "Product bundle" from the product data section.
- Set the price.
- Go to "Bundled Products" then add at least two simple products.
- Publish.

**Create a bundle with virtual products**

- Go to Products > Add New.
- Select "Product bundle" from the product data section.
- Set the price.
- Go to "Bundled Products" then add at least two simple products.
- Publish.

**Create a bundle with mixed products (simple and virtual)**

- Go to Products > Add New.
- Select "Product bundle" from the product data section.
- Set the price.
- Go to "Bundled Products" and add both a simple product and a virtual product.
- Publish.

**Test cases**

- Bundle - Simple products
    - Product Page
        - [ ] ❌ Without Tax
        - [ ] ❌ With Tax (Customer Shipping Address)
        - [ ] ❌ With Tax (Customer Billing Address)
        - It doesn't work because it asks for a shipping address but there's no shipping section in the payment sheet. Looks like a bug in the plugin. We could disable ECE entirely for this product type in the product page.
    - Shortcode - Cart
        - [x] Without Tax
        - [x] With Tax (Customer Shipping Address)
        - [ ] ❌ With Tax (Customer Billing Address) – **Physical Product With Tax Error**
    - Shortcode - Checkout
        - [x] Without Tax
    - Blocks - Cart
        - [x] Without Tax
        - [x] With Tax (Customer Shipping Address)
        - [ ] ❌ With Tax (Customer Billing Address) – **Physical Product With Tax Error**
    - Blocks - Checkout
        - [x] Without Tax
- Bundle - virtual products
    - Product Page
        - [x] Without Tax
        - [ ] ❌ With Tax (Customer Shipping Address)
        - [ ] ❌ With Tax (Customer Billing Address) – **Physical Product With Tax Error**
    - Shortcode - Cart
        - [x] Without Tax
        - [x] With Tax (Customer Shipping Address)
        - [ ] ❌ With Tax (Customer Billing Address) – **Physical Product With Tax Error**
    - Shortcode - Checkout
        - [x] Without Tax
    - Blocks - Cart
        - [x] Without Tax
    - Blocks - Checkout
        - [x] Without Tax
- Bundle - mixed products
    - Product Page
        - [ ] ❌ Without Tax
        - [ ] ❌ With Tax (Customer Shipping Address)
        - [ ] ❌ With Tax (Customer Billing Address)
    - Shortcode - Cart
        - [x] Without Tax
        - [x] With Tax (Customer Shipping Address)
        - [ ] ❌ With Tax (Customer Billing Address) – **Physical Product With Tax Error**
    - Shortcode - Checkout
        - [x] Without Tax
    - Blocks - Cart
        - [x] Without Tax
    - Blocks - Checkout
        - [x] Without Tax
</details>

### Product type: `composite`

<details>
<summary>Testing steps</summary>

In order to test this product type you'll need to install the [Composite Products](https://woocommerce.com/products/composite-products/) plugin. You can download it from [here](https://woocommerce.com/my-account/downloads/).


**Create a composite product with simple products**

- Go to Products > Add New.
- Select "Composite product" from the product data section.
- Set the price.
- Go to the "Components" tab
- Add a couple of components, each containing two simple products.
- Publish.

**Note:** No need to test the ECE on the product page for this product type as it's currently not supported in that page.

**Test cases**

- Composite - Simple products
  - Blocks - Cart
    - [x] Without Tax
  - Blocks - Checkout
    - [x] Without Tax
</details>

### Product type: `mix-and-match`

<details>
<summary>Testing steps</summary>

In order to test this product type you'll need to install the [Mix and Match Products](https://woocommerce.com/products/woocommerce-mix-and-match-products/) plugin. You can download it from [here](https://woocommerce.com/my-account/downloads/).


**Create a "Mix and Match" product with simple products**

- Go to Products > Add New.
- Select "Mix and Match product" from the product data section.
- Set the price.
- Go to the "Mix and Match" tab.
- Select a couple simple products.
- Publish.

**Note:** No need to test the ECE on the product page for this product type as it's currently not supported in that page.

**Test cases**

- Mix and Match - Simple products
  - Blocks - Cart
    - [x] Without Tax
  - Blocks - Checkout
    - [x] Without Tax
</details>


--- 
#### Error types
- **Virtual Product With Tax Error (Shipping)**
  - We can't determine the tax amount based on a shipping address because we don't have a shipping address.
  - There's no section to select a shipping address in the payment sheet.
- **Virtual Product With Tax Error (Billing)**
  - We can't get the billing information from the payment method before displaying the payment sheet.
  - Tax amount is calculated using the store's country or whatever `Default customer location`" option is set to.
  - Payment is done but the amount shown was different.
  - There's no section to select a shipping address in the payment sheet so it seems to use the billing address instead.
- **Physical Product With Tax Error**
  - Fails to display the tax amount based on the billing address since we can not determine the selected payment method before showing the payment sheet.
---

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8987` or branch name `as-8871-ece-on-supported-product-types` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=as-8871-ece-on-supported-product-types) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 33d4f171fe350f4387396e2f5ef334dcf196f7d8
- Build time: 2024-07-08 20:48:32 UTC

_Note: the build is updated when a new commit is pushed to this PR._

asumaran committed:
Fix total amount for virtual products

asumaran committed:
Remove duplicated changelog entry

asumaran committed:
Add ECE support for variable products

asumaran committed:
Fix button not getting blocked when changing a product attribute

asumaran committed:
Add TODO to improve button updates

asumaran committed:
Add ECE support for payments that require authentication first

asumaran committed:
Remove unnecessary hook

Tested multiples scenarios and was never necessary

asumaran committed:
Fix login confirmation for the cart page

asumaran committed:
Set WC session if no session is found

Fixes an issue where ECE fails if the cookies are removed before
refresh.

asumaran committed:
Catch possible erros when creating the order

Fixes an issue where the UI is stuck if the request to create the order
returns an invalid response.

asumaran committed:
Add the fraud prevention token to the product page

asumaran committed:
Optimize input handling with debounce

asumaran committed:
Adjust selector to make it work with bundles

The bundle plugin has multiple .qty inputs, and the one we use to determine the quantity is the last input, which is of the number type.

asumaran committed:
Merge branch 'develop' into as-8871-ece-on-supported-product-types

# Conflicts:
#	client/express-checkout/index.js
#	client/express-checkout/utils/index.ts

asumaran committed:
Reset paymentAborted flag every time the ECE is instantiated

cesarcosta99 commented on client/express-checkout/event-handlers.js:
```diff
@@ -101,7 +101,10 @@ export const onConfirmHandler = async (
-		return abortPayment( event, e.message );
+		return abortPayment(
+			event,
+			e.message ?? 'There was a problem creating the order'
```
Is this error displayed to the shopper? If so, we may want to localize this literal string for translation.

asumaran replied:
>Is this error displayed to the shopper?

Yes it is. That message will displayed if an error happens during the `confirm` event. e.g. if the cart is empty. https://github.com/Automattic/woocommerce-payments/blob/022ef8dc76754a1fd4fce04249f5d678cfb1fcb4/includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php#L49-L51

>we may want to localize this literal string for translation.

done! https://github.com/Automattic/woocommerce-payments/pull/8987/commits/71e4d25100cac2ff1d8c25d3ca19f51798052608

cesarcosta99 commented on client/express-checkout/index.js:
```diff
@@ -30,6 +32,7 @@ jQuery( ( $ ) => {
 	}
 
 	const publishableKey = wcpayExpressCheckoutParams.stripe.publishableKey;
+	const quantityInoutSelector = '.quantity .qty[type=number]';
```
I guess this is a typo? If so, we need to replace it in the references as well.

```suggestion
	const quantityInputSelector = '.quantity .qty[type=number]';
```

asumaran replied:
👀 Good catch! Fixed https://github.com/Automattic/woocommerce-payments/pull/8987/commits/17c03e6ae2d03d3e74080ca40d23f05543a3fcbb.

cesarcosta99 commented on client/express-checkout/index.js:
```diff
@@ -250,9 +255,47 @@ jQuery( ( $ ) => {
+					}
+
+					// TODO: This scenario.
+					// For PRBs wcPayECEError (paymentRequestError) is never re-assigned so it might be not working properly.
```
Are we addressing this TODO in this PR? If not, should we create a new issue to test this out?

asumaran replied:
This was already addressed in https://github.com/Automattic/woocommerce-payments/pull/8987/commits/1544d9b4b1992996158ade9d1187730caa009714

asumaran replied:
Prepared the error message for localization here https://github.com/Automattic/woocommerce-payments/pull/8987/commits/108499c5b423c521c5351a6135747781bf23e748

cesarcosta99 commented on client/express-checkout/index.js:
```diff
@@ -250,9 +255,47 @@ jQuery( ( $ ) => {
+					// TODO: This scenario.
+					// For PRBs wcPayECEError (paymentRequestError) is never re-assigned so it might be not working properly.
+					if ( wcPayECEError.length > 0 ) {
+						window.alert( wcPayECEError );
```
Should we create an issue to replace this and above window alerts with a built-in WooCommerce error? Displaying errors like that is not the best user experience we can offer, especially `wcPayECEError` being an array that can potentially carry multiple errors.

asumaran replied:
>Should we create an issue to replace this and above window alerts with a built-in WooCommerce error? Displaying errors like that is not the best user experience we can offer

Currently, this is how PRBs function. While I agree that improvements can be made, it would be best to address those in a separate issue. What do you think, @bborman22? Is it worth exploring this further?


>especially wcPayECEError being an array that can potentially carry multiple errors.

I've already updated the code to treat `wcPayECEError` as string but improved it in 
https://github.com/Automattic/woocommerce-payments/pull/8987/commits/8ebe1adf70249aabaf599fd87e65f8f4e6c8eb25

bborman22 replied:
I would agree that we should fix this in a separate issue. We should create one and make sure it has the focus of checkout payments and type of enhancement so it gets triaged properly.

cesarcosta99 replied:
@asumaran, could you create an issue for this?

asumaran replied:
Issue created https://github.com/Automattic/woocommerce-payments/issues/9069

cesarcosta99 commented on client/express-checkout/index.js:
```diff
@@ -353,44 +401,127 @@ jQuery( ( $ ) => {
+							 * we need to re init the payment request button to ensure the shipping
+							 * options are refetched. If the customer didn't abort the payment request,
+							 * and the product's shipping status is consistent,
+							 * we can simply update the payment request button with the new total and display items.
```
We may want to replace references of "payment request" with "express checkout" here.

asumaran replied:
done in https://github.com/Automattic/woocommerce-payments/pull/8987/commits/329852784e122b10abedaba2e1534bf93fb5f0e1 and https://github.com/Automattic/woocommerce-payments/pull/8987/commits/1596331f637212a6f09e9124ca939d4d01a3ba36

asumaran committed:
Handle failure if fetching the selected product data fails

cesarcosta99 commented on client/express-checkout/index.js:
```diff
@@ -353,44 +401,127 @@ jQuery( ( $ ) => {
+								}
+
+								// TODO: Cache the `needs_shipping` value of the last response so
+								// we re-init only when necessary, that is when `needs_shipping` has changed.
```
Are we addressing this TODO in this PR? If not, should we create a new issue to address this?

asumaran replied:
Already addressed. The TODO was removed.

cesarcosta99 commented on includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php:
```diff
@@ -35,10 +38,15 @@ public function __construct( WC_Payments_Express_Checkout_Button_Helper $express
 		add_action( 'wc_ajax_wcpay_update_shipping_method', [ $this, 'ajax_update_shipping_method' ] );
+		add_action( 'wc_ajax_wcpay_get_selected_product_data', [ $this, 'ajax_get_selected_product_data' ] );
+		add_filter( 'woocommerce_login_redirect', [ $this, 'get_login_redirect_url' ], 10, 3 );
+		add_filter( 'woocommerce_registration_redirect', [ $this, 'get_login_redirect_url' ], 10, 3 );
```
Given the name of this class, I believe we should only hook AJAX calls here. We may want to move the other hooks to [this section](https://github.com/Automattic/woocommerce-payments/blob/05e7d8629eefb351b15a57e9af609781e5fa21e9/includes/express-checkout/class-wc-payments-express-checkout-button-handler.php#L90-L91).

asumaran replied:
Good call. Done in b8b26e6765921fa885667e7562f6fdc4715514b3

cesarcosta99 commented on includes/express-checkout/class-wc-payments-express-checkout-button-handler.php:
```diff
@@ -109,6 +109,99 @@ public function get_button_settings() {
+		$redirect_url = add_query_arg(
+			[
+				'_wpnonce'                           => wp_create_nonce( 'wcpay-set-redirect-url' ),
+				'wcpay_payment_request_redirect_url' => rawurlencode( home_url( add_query_arg( [] ) ) ),
```
To refresh the wording, we may want to rename `wcpay_payment_request_redirect_url` to `wcpay_express_checkout_redirect_url`. If you agree, we need to replace the references in this file and in the `get_login_redirect_url` and `handle_payment_request_redirect` functions that are likely getting moved to this file as well, per the comment above.

asumaran replied:
Done in https://github.com/Automattic/woocommerce-payments/commit/80e18ba28aa4e80e31c78da8d4a589272c889e4d

asumaran committed:
Add changelog entry

cesarcosta99 commented on includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php:
```diff
@@ -257,4 +474,41 @@ public function ajax_empty_cart() {
+	/**
+	 * Handles payment request redirect when the redirect dialog "Continue" button is clicked.
+	 */
+	public function handle_payment_request_redirect() {
```
We can rename this function to `handle_express_checkout_redirect`.

asumaran replied:
Good call. done in https://github.com/Automattic/woocommerce-payments/pull/8987/commits/80e18ba28aa4e80e31c78da8d4a589272c889e4d

cesarcosta99 commented on includes/fraud-prevention/class-fraud-prevention-service.php:
```diff
@@ -88,7 +88,7 @@ public static function maybe_append_fraud_prevention_token() {
 
 		// Don't add the token if the user isn't on the cart or checkout page.
 		// Checking the cart page too because the user can pay quickly via the payment buttons on that page.
-		if ( ! is_checkout() && ! is_cart() ) {
+		if ( ! is_checkout() && ! is_cart() && ! is_product() ) {
```
Nice catch! Would be nice to add a comment here providing context as to why we're skipping the fraud prevention token in the product page as well.

asumaran replied:
We are not skipping it. Here we are adding the Fraud Prevention Token to the product page to make it work. Without it, the order wouldn’t process. This change addresses issue #8981.

cesarcosta99 replied:
Sorry, I misread the comment. But what I mean by that is that we should add a comment mentioning why we're adding the fraud prevention token to the product page as well (similar to the cart comment). Something like:

```diff
+ // Checking the product page too because the user can pay through express checkout (ECE/PRB) on that page.
```

This is just to provide context to folks who are not aware of why the token is needed in there.

asumaran replied:
Note that this token is not exclusive to ECE or PRBs. While express checkouts use it, it is intended for use on any page that submits an order for processing. Done in https://github.com/Automattic/woocommerce-payments/pull/8987/commits/bcf8bfa43c1e827836b15de524aeb24c05db3f71

cesarcosta99 commented on changelog/add-8870-ece-for-shortcode-cart-and-checkout:
```diff
@@ -1,4 +0,0 @@
-Significance: minor
-Type: add
-
-Add support for Express Checkout Element on shortcode Cart and Checkout pages.
```
Can you expand on why we're removing this change log?

asumaran replied:
This file was added by mistake in https://github.com/Automattic/woocommerce-payments/pull/8937 because the PR was based on the same branch @reykjalin used for his own [PR](https://github.com/Automattic/woocommerce-payments/pull/8914). 

Context: p1718967918819489-slack-C01BZUL57SQ

asumaran replied:
Nevermind, it was removed here https://github.com/Automattic/woocommerce-payments/pull/9054

asumaran committed:
Attempt to fix PHPCS errors

cesarcosta99 reviewed (changes requested):
Overall the code looks good, nice work! I've made some suggestions that involves renaming, refactoring and moving things around.

Since there's a potential refactoring to be done and the testing is very time consuming, I'm going to hold off the testing until you request a new review.

asumaran committed:
Merge branch 'develop' into as-8871-ece-on-supported-product-types

# Conflicts:
#	client/express-checkout/event-handlers.js
#	includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php

asumaran committed:
Silence psalm error

asumaran committed:
Restore code

asumaran committed:
Fix failing JS test

asumaran commented:
@rafaelzaleski You were randomly selected to review, Please focus on testing `booking`, `bundles`, `composite`, and `mix-and-match`. Let me know if you have any questions!

I will keep improving the PR descriptions but it should be ready for review.

asumaran committed:
Localize error message during the confirm event

asumaran committed:
Typo

asumaran committed:
Localize error message

asumaran committed:
Treat wcPayECEError as string

asumaran committed:
Changed references from 'payment request' to 'express checkout'

asumaran committed:
Improve the displayLoginConfirmation

- Updated references from PRB to ECE.
- Updated to handle ECE payment methods.

asumaran committed:
Rename for better legibility

asumaran committed:
Move hooks to a proper place

asumaran committed:
Merge branch 'develop' into as-8871-ece-on-supported-product-types

rafaelzaleski commented:
@asumaran, the checkboxes and the ❌ symbol’s meanings are unclear. Could you also explain why we’re testing without taxes for some extensions and with taxes for others?

@bborman22, this PR seems overloaded with test cases. Maybe we should have divided it based on the different extensions from the beginning. Could someone from the team help test WooCommerce Bookings?

I’ll take care of testing the product bundles, mix and match, and composites.

@asumaran, who is assigned to test subscriptions and the other product types since you asked me to focus on bookings, bundles, mix and match and composite?

asumaran commented:
> the checkboxes and the ❌ symbol’s meanings are unclear.

@rafaelzaleski I have tested all the test cases in the list. If the checkbox is marked, that means the test passed.

The unchecked items (with the ❌ mark) didn't pass and are currently failing for ECE and PRB. I put the reason each test failed next to them. The details for each error are at the end of the PR description.

The reviewer should review only the test that passed to confirm that ECE supports the product type and to save time reviewing most of the cases that we know will surely fail. As you can see, most of the failed cases are related to Tax rates and are being investigated [here](https://github.com/Automattic/woocommerce-payments/issues/8373) 

Given that the tests that didn't pass won't work on PRBs as well we should work on disabling ECE for those particular cases or figuring out a solution in a separate issue.

>Could you also explain why we’re testing without taxes for some extensions and with taxes for others?

Given the number of test cases for each product type, I only tested some possible scenarios on all the pages where the ECE is shown. Focusing on one case at least on one page gives us an idea of how it will behave on other pages. I've only tested the last product types without tax to confirm that the ECE works for the given product type. I was planning to add more test cases, but I didn't want to hold off on reviewing this PR because the core logic has not changed since days ago. 

>who is assigned to test subscriptions and the other product types since you asked me to focus on bookings, bundles, mix and match and composite?

@cesarcosta99 should be responsible for reviewing the initial product types up to and including the subscription_variation product type.

>this PR seems overloaded with test cases. Maybe we should have divided it based on the different extensions from the beginning.

The problem is that most of the code applies to all product types. Dividing the work might have led to the same scenario as before, where @reykjalin and I ended up duplicating code.

rafaelzaleski commented:
Nice! Thanks for the detailed reply, @asumaran. I’m starting the tests now!

asumaran committed:
Explain in the comments why the token is added to the product page

asumaran committed:
Merge branch 'develop' into as-8871-ece-on-supported-product-types

cesarcosta99 commented on client/express-checkout/utils/index.ts:
```diff
@@ -111,6 +119,52 @@ export const getErrorMessageFromNotice = ( notice: string ) => {
+		link: 'Link',
+	};
+	let message =
+		'To complete your transaction with **the selected payment method**, you must log in or create an account with our site.';
```
Is there a reason for not using the `message` that comes in `loginConfirmation`? If we want to move the message to the frontend, we may want to localize it here and remove it from the backend.

asumaran replied:
No reason. done here [fe82088](https://github.com/Automattic/woocommerce-payments/pull/8987/commits/fe820880e405e0e285f5a23ab8563de9e7bf1ceb)

rafaelzaleski reviewed (changes requested):
@asumaran, I’m still working through the many test cases, but one issue stands out: we’re rendering ECE payment methods in scenarios where they’re known to fail or charge incorrectly, especially concerning taxes. 

We should verify the cart and store settings to enhance the shopper experience, ensuring we only display payment methods when we’re confident the payment can be processed successfully.

The other feedback that I can give based on what I've already tested:

**Bundle - Virtual Products**

- Product page
    - Worked for taxes calculated based on billing address but failed when based on shipping address. When based on shipping address, it adds taxes based in the billing address from GPay.
    - Also, when calculated based on billing, it shows the incorrect amount in the GPay prompt but charges correctly. For instance, it shows the value without taxes but charges with taxes, which was the correct amount.

**Bundle - Mixed Products**

- Product page: the buttons are rendering in the page (if they don't work we shouldn't render them), but it's not possible to finish the payment:

> The following problems were found: The payment request button is not supported in United States (US) because some required fields couldn't be verified. Please proceed to the checkout page and try again. Shipping First name is a required field. Shipping Last name is a required field. Shipping Country / Region is a required field. Shipping Street address is a required field. Shipping Town / City is a required field. Shipping State is a required field. Shipping ZIP Code is a required field. Please enter an address to continue.
>

asumaran commented:
>we’re rendering ECE payment methods in scenarios where they’re known to fail or charge incorrectly, especially concerning taxes.

@rafaelzaleski That's correct. I believe hiding ECE for those cases deserves its own issue in case we don't find a solution. I didn't include it in this PR because the testing could have been nearly endless 😀 (or maybe not, but I thought on handling it separately to simplify the testing process).

>Bundle - Virtual Products
>Worked for taxes calculated based on billing address.

You sure? You might get charged the correct amount but what was shown to the user was incorrect.

<img width="293" alt="Screenshot 2024-07-05 at 2 04 18 PM" src="https://github.com/Automattic/woocommerce-payments/assets/1025173/de71bf13-2b91-486c-bc20-c747723a4f86">

Assuming:
USA TAX is set to 11%
Perú Tax is set to 35%

Check this example using a Bundle of virtual products. It shows `$3.74` as tax amount. Product is `34 + 34*0.11 = $37.74`

However the user is charged `$45.90` 
<img width="493" alt="Screenshot 2024-07-05 at 2 05 22 PM" src="https://github.com/Automattic/woocommerce-payments/assets/1025173/ea5c31f4-f85f-4185-8bc4-1219e18fdebe">

Although the end result is correct, to me that's incorrect and I marked them as failed. We can't show the user an incorrect amount then charge another.

>Bundle - Mixed Products
>Product page: the buttons are rendering in the page (if they don't work we shouldn't render them), but it's not possible to finish the payment:

I agree. As mentioned above I think we should handle it in a separate issue.

rafaelzaleski commented:
The only inconsistency I found when testing `product bundles`, `composite`, and `mix-and-match` was:

Bundle - virtual products
- Product Page
-- Does not work when taxes are enabled (based on shipping or billing). 

I'll test Bookings on Monday.

rafaelzaleski commented on client/express-checkout/utils/index.ts:
```diff
@@ -111,6 +119,52 @@ export const getErrorMessageFromNotice = ( notice: string ) => {
+		link: 'Link',
+	};
+	let message =
+		'To complete your transaction with **the selected payment method**, you must log in or create an account with our site.';
```
Isn’t this message already a property of `loginConfirmation`? Why do we need to set it again? 🤔 

If there is a need to repeat it here, should we add support for internationalization?

asumaran replied:
Good catch. I've fixed it here [fe82088](https://github.com/Automattic/woocommerce-payments/pull/8987/commits/fe820880e405e0e285f5a23ab8563de9e7bf1ceb)

rafaelzaleski commented on includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php:
```diff
@@ -212,6 +216,197 @@ public function ajax_update_shipping_method() {
+
+		// Normally there should be a single tax, but `calc_tax` returns an array, let's use it.
+		return WC_Tax::calc_tax( $price, $rates, false );
+	}
```
These methods are already present in the button helper class. Additionally, since they aren’t Ajax handlers, I don’t think they belong here.

asumaran replied:
Good catch! Addressed in https://github.com/Automattic/woocommerce-payments/pull/8987/commits/612ea27d9ee453624acbf0f5d2aa4023da7281af

asumaran replied:
Had to make `get_taxes_like_cart` public so it can be used from the outside. I hope that's fine.  https://github.com/Automattic/woocommerce-payments/pull/8987/commits/33d4f171fe350f4387396e2f5ef334dcf196f7d8

cesarcosta99 committed:
Merge branch 'develop' into as-8871-ece-on-supported-product-types

cesarcosta99 commented on changelog/as-8871-ece-on-supported-product-types:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: add
+
+Add ECE support for multiple product types
```
Nit:

```suggestion
Add ECE support for multiple product types.
```

asumaran replied:
done [c8b749b](https://github.com/Automattic/woocommerce-payments/pull/8987/commits/c8b749b7982cc2dfa36126468b805c42fc824772)

rafaelzaleski reviewed (changes requested):
Overall, the code looks good. 

I noticed extensive use of jQuery for simple selectors, likely because the code was copied from PRBs. I won’t extend this review further on that point, as we have a dedicated issue to reduce jQuery usage in the ECE implementation.

I also noticed code related to `WooCommerce Deposits`, even though these aren’t listed in `WC_Payments_Express_Checkout_Button_Helper->supported_product_types`. While this is in the scope of issue #8871, I’d rather not complicate this PR further. Could you handle this in a separate PR? Please either remove the code related to deposits if we aren’t supporting them, or ensure all necessary support is included before closing issue #8871.

Please focus on my comments as I believe they are blockers.

Thanks for all your hard work! 🙏

cesarcosta99 reviewed (changes requested):
I appreciate the updates you did, great work! Just one last thing that came through in the latest changes.

All tests that I run passed 🎉 

- Virtual Product ✅ 
- Physical Product ✅ 
    - Without Tax
    - With Tax (Customer Shipping Address)
- Ensure the ECE button renders with a variable product ✅ 
- Product Type: `variation` ✅ 
    - Virtual Products
    - Physical Products
        - Without Tax
        - With Tax (Customer Shipping Address)
    - Mixed Products (Virtual & Physical)
        - Virtual Variation
        - Physical Variation
            - Without Tax
            - With Tax (Customer Shipping Address)
- Product Type: `subscription` ✅ 
    - Virtual
    - Physical
        - Without Tax
        - With Tax (Customer Shipping Address)
- Ensure the ECE button renders with a variable subscription ✅ 
- Product Type: `subscription_variation` ✅ 
    - Virtual
    - Physical
        - Without Tax
        - With Tax (Customer Shipping Address)
    - Mixed (Virtual & Physical)
        - Virtual Variation
        - Physical Variation
            - Without Tax
            - With Tax (Customer Shipping Address)


>Given that the tests that didn't pass won't work on PRBs as well we should work on disabling ECE for those particular cases or figuring out a solution in a separate issue.
> – https://github.com/Automattic/woocommerce-payments/pull/8987#issuecomment-2209658359

> We should verify the cart and store settings to enhance the shopper experience, ensuring we only display payment methods when we’re confident the payment can be processed successfully.
> – https://github.com/Automattic/woocommerce-payments/pull/8987#pullrequestreview-2161101374

To touch on the above, it looks like this tax thing is already an issue present in PRBs (#8373). Since this is not a regression, IMO, we should use that spike to address this. The testing in this PR is extremely extensive already, and during that spike we can take advantage of the test instructions here to ensure everything is working as expected.

Another aspect we should consider when revisiting this is to confirm if it really makes sense to tax virtual products. Maybe we should consult the team working on the tax feature and/or tax experts to be sure we're applying taxes correctly. Based on whether tax or not virtual products, we can decide to either hide the ECE button, or fix the application of the tax for the (virtual) products where it's buggy. cc @bborman22

rafaelzaleski commented:
I’ve finished testing WC Bookings, and it performs as described in the testing instructions.

asumaran committed:
Typo

asumaran committed:
Use login confirmation message that comes from the backend.

asumaran committed:
Remove duplicated methods and re-use them from the Button Helper

asumaran committed:
Remove unnecessary psalm base

asumaran commented:
>I also noticed code related to WooCommerce Deposits, even though these aren’t listed in WC_Payments_Express_Checkout_Button_Helper->supported_product_types. While this is in the scope of issue https://github.com/Automattic/woocommerce-payments/issues/8871, I’d rather not complicate this PR further. Could you handle this in a separate PR?

@rafaelzaleski Deposits wasn't part of the list on the issue and I missed it completely. I'll create an issue to make sure we are supporting Deposits as well.

--

Issue created https://github.com/Automattic/woocommerce-payments/issues/9067

asumaran committed:
Make get_taxes_like_cart method public

This is to make it possible it can be used from class-wc-payments-express-checkout-ajax-handler.php

asumaran commented:
>The only inconsistency I found when testing product bundles, composite, and mix-and-match was:
>
>Bundle - virtual products
>
>Product Page
>-- Does not work when taxes are enabled (based on shipping or billing).

@rafaelzaleski You are correct. I must have copied it incorrectly [from my list](https://large-cause-57b.notion.site/8871-Add-ECE-support-for-multiple-product-types-9f6ddb6bc5674f289aca6179870be7a3?pvs=4), which correctly shows that bundles with virtual products fail with taxes. I've updated the PR description to reflect the correct outcome.

asumaran commented:
Thanks for your feedback, @cesarcosta99 @rafaelzaleski. I think I addressed all your comments and code suggestions, also created issues where needed. If the PR looks good to you, would you mind approving it?

rafaelzaleski reviewed (approved):
LGTM! Thank you for making the changes and creating the issue for WC Deposits.

cesarcosta99 reviewed (approved):
LGTM! Thank you for addressing all of our feedback and creating the follow-up issues 💯
